### ISSUE TITLE
Performance Optimizations

### ISSUE THREAD

**Issue by brettdewoody** [Munter/netlify-plugin-subfont#100]
I'd like to start a discussion around potential performance optimizations. This stems from my experience with SubFont where my build time has gone from ~4 minutes to ~20+ minutes when using the SubFont plugin. 

My site is built with Eleventy, and has ~140 pages.

Would there be any performance improvement from explicitly defining all the pages as `entryPoints` instead of relying on `recursive`?

What other potential performance improvements could we explore?

**Comment by Munter** [comment]
If I recall correctly, the slowest parts of subfont is the graph traversal, including the parsing and AST creation for each asset type, and then the quite extensive analysis of traversing each text node to calculate its specific font properties (with inherited properties).

Instantiating the graph could potentially become faster with a few more entry points. But I assume you'd only gain a few milliseconds, since most pages quite rapidly queue up more relations to be populated by the graph traversal, than can be reasonably processed in parallel at any given time.

There's a central population queue, which is populated with a 100 concurrent assets at any one time. This is enough to not have much I/O downtime while populating, once the queue is at a 100+ items, so I don't think there is much to gain here.

We are not able to parallelize graph population work in new threads with webworkers, because the architecture requires everything to be in the same memory space. This makes our primary bottleneck the CPU during population.

While running analysis of font style applications on all text nodes and non-text nodes where font styles matter, there's a lot 
 of traversals going on. We've attempted to reduce the amount of work with some clever heuristics, but we haven't established any explicit performance focused benchmarks and worked to improve in this area yet.
 
@papandreou do you have a better recollection of where the bottlenecks might be here, and if there are any low hanging fruits?

**Comment by papandreou** [comment]
IIRC the slow part is the tracing, especially when you hit a bad case with many different media/supports queries containing font-related CSS properties. Animations or cudtom properties (CSS variables) and combinations thereof also add to it.

But with 140 pages the loading probably also adds up 😬

I think we’ll need to see a profile of the execution or be able to run the build ourselves to be able to tell what’s slow. Are you able to share your repo?

**Comment by brettdewoody** [comment]
In my case, I did some experiments and discovered that by merely preloading the fonts I got nearly all the performance benefits. So in the end I removed `subfont` and just preloaded the fonts. With `subfont` the performance was slightly better, but the build time increase from 3 minutes to 20+ minutes didn't make it worth it (I'd exceed my build minutes on Netlify very quickly). 

So, perhaps an option to ignore subsetting and just preload or inline fonts would be a good tradeoff?

**Comment by papandreou** [comment]
Yeah, we’ve already thought about splitting up the tool so the self-hosting feature could be used on its own. It’s a good idea to also separate out the preloading like that, let’s make a note of it.

I still think it’d also be interesting to look for easy wins in the tracing code, so I’d still be happy if you could share your project or a tarball of the assets that you were running subfont on 🤗

**Comment by kotgakapainor** [comment]
I just started using subfont for netlify and at first I thought the plugin did break/freeze my build. At least add a warning to the readme that subfont may quadruple the build time.

kind regards
